File: lib/models/banner.dart
class Banner {
  final String id;
  final String title;
  final String subtitle;
  final String imageUrl;
  final String actionUrl;
  final String actionText;
  final int priority;
  final bool isActive;
  final DateTime startDate;
  final DateTime? endDate;
  final String targetAudience;
  final String platform;
  final DateTime createdAt;
  final DateTime updatedAt;
  
  Banner({
    required this.id,
    required this.title,
    required this.subtitle,
    required this.imageUrl,
    required this.actionUrl,
    required this.actionText,
    required this.priority,
    required this.isActive,
    required this.startDate,
    this.endDate,
    required this.targetAudience,
    required this.platform,
    required this.createdAt,
    required this.updatedAt,
  });
  
  factory Banner.fromJson(Map<String, dynamic> json) {
    return Banner(
      id: json['id'] as String,
      title: json['title'] as String? ?? '',
      subtitle: json['subtitle'] as String? ?? '',
      imageUrl: json['imageUrl'] as String,
      actionUrl: json['actionUrl'] as String? ?? '',
      actionText: json['actionText'] as String? ?? 'Learn More',
      priority: json['priority'] as int? ?? 0,
      isActive: json['isActive'] as bool? ?? true,
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'].toString())
          : DateTime.now(),
      endDate: json['endDate'] != null
          ? DateTime.parse(json['endDate'].toString())
          : null,
      targetAudience: json['targetAudience'] as String? ?? 'all',
      platform: json['platform'] as String? ?? 'all',
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'].toString())
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'].toString())
          : DateTime.now(),
    );
  }
  
  bool isCurrentlyActive() {
    if (!isActive) return false;
    final now = DateTime.now();
    if (now.isBefore(startDate)) return false;
    if (endDate != null && now.isAfter(endDate!)) return false;
    return true;
  }
}

File: server/src/models/Banner.js
/**
 * Banner Model
 * Dynamic promotional banners that can be updated without app rebuilds
 */
export class Banner {
  constructor(data) {
    this.id = data.id;
    this.title = data.title || '';
    this.subtitle = data.subtitle || '';
    this.imageUrl = data.imageUrl || '';
    this.actionUrl = data.actionUrl || '';
    this.actionText = data.actionText || 'Learn More';
    this.priority = data.priority || 0; // Higher priority = shown first
    this.isActive = data.isActive ?? true;
    this.startDate = data.startDate ? new Date(data.startDate) : new Date();
    this.endDate = data.endDate ? new Date(data.endDate) : null;
    this.targetAudience = data.targetAudience || 'all'; // 'all', 'new_users', 'premium'
    this.platform = data.platform || 'all'; // 'all', 'ios', 'android'
    this.createdAt = data.createdAt || new Date();
    this.updatedAt = data.updatedAt || new Date();
  }

  toFirestore() {
    return {
      title: this.title,
      subtitle: this.subtitle,
      imageUrl: this.imageUrl,
      actionUrl: this.actionUrl,
      actionText: this.actionText,
      priority: this.priority,
      isActive: this.isActive,
      startDate: this.startDate,
      endDate: this.endDate,
      targetAudience: this.targetAudience,
      platform: this.platform,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }

  static fromFirestore(doc) {
    const data = doc.data();
    return new Banner({
      id: doc.id,
      ...data,
    });
  }

  isCurrentlyActive() {
    if (!this.isActive) return false;
    const now = new Date();
    if (now < this.startDate) return false;
    if (this.endDate && now > this.endDate) return false;
    return true;
  }
}

